where() fails after renaming columns on table read from parquet file

If I read a table from Parquet, then rename one of its columns with a view(), then run a where() on the renamed column, I get an exception while compiling the filter. However the code works fine if the view() and where() are run in separate commands in the console.

DH Version 0.14.1

Reproducing code:

```
SIMULATED_STREAKS_PATH = '/dh/data/shared/baseball/simulation/bts_simulation_results_2021.parquet'
contestants_with_streaks = io.deephaven.parquet.table.ParquetTools.readTable(SIMULATED_STREAKS_PATH)
tst = contestants_with_streaks.view(
        'Batter=Pick1Batter'
    )
println tst.getDefinition().getColumnNames()
// this works fine if you run it separately, but not when you copy/paste
// it at the same time as the .view() above
tst2 = tst.where('!isNull(Batter)')
```


Exception:

```
r-Scheduler-Serial-1 | i.d.s.s.SessionState      | Internal Error 'c020825d-73f1-4a6d-9b05-05877c59386f' io.deephaven.engine.table.impl.select.FormulaCompilationException: Formula compilation error for: !isNull(Batter)
        at io.deephaven.engine.table.impl.select.AbstractConditionFilter.init(AbstractConditionFilter.java:200)
        at io.deephaven.engine.table.impl.QueryTable.lambda$whereInternal$28(QueryTable.java:1023)
        at io.deephaven.engine.table.impl.perf.QueryPerformanceRecorder.withNugget(QueryPerformanceRecorder.java:518)
        at io.deephaven.engine.table.impl.QueryTable.whereInternal(QueryTable.java:993)
        at io.deephaven.engine.table.impl.QueryTable.where(QueryTable.java:989)
        at io.deephaven.engine.table.impl.TableWithDefaults.where(TableWithDefaults.java:212)
        at io.deephaven.engine.table.impl.DeferredViewTable.getResultTableWithWhere(DeferredViewTable.java:103)
        at io.deephaven.engine.table.impl.DeferredViewTable.where(DeferredViewTable.java:73)
        at io.deephaven.engine.table.impl.TableWithDefaults.where(TableWithDefaults.java:218)
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:566)
        at org.codehaus.groovy.runtime.callsite.PlainObjectMetaMethodSite.doInvoke(PlainObjectMetaMethodSite.java:43)
        at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite$PojoCachedMethodSite.invoke(PojoMetaMethodSite.java:191)
        at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite.call(PojoMetaMethodSite.java:56)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:139)
        at io.deephaven.dynamic.Script_32.run(Script_32.groovy:45)
        at groovy.lang.GroovyShell.evaluate(GroovyShell.java:427)
        at groovy.lang.GroovyShell.evaluate(GroovyShell.java:461)
        at groovy.lang.GroovyShell.evaluate(GroovyShell.java:436)
        at io.deephaven.engine.util.GroovyDeephavenSession.evaluateCommand(GroovyDeephavenSession.java:228)
        at io.deephaven.engine.util.GroovyDeephavenSession.lambda$evaluate$2(GroovyDeephavenSession.java:251)
        at io.deephaven.util.locks.FunctionalLock.doLockedInterruptibly(FunctionalLock.java:49)
        at io.deephaven.engine.util.GroovyDeephavenSession.evaluate(GroovyDeephavenSession.java:251)
        at io.deephaven.engine.util.AbstractScriptSession.evaluateScript(AbstractScriptSession.java:166)
        at io.deephaven.engine.util.DelegatingScriptSession.evaluateScript(DelegatingScriptSession.java:81)
        at io.deephaven.engine.util.ScriptSession.evaluateScript(ScriptSession.java:106)
        at io.deephaven.server.console.ConsoleServiceGrpcImpl.lambda$executeCommand$8(ConsoleServiceGrpcImpl.java:188)
        at io.deephaven.server.session.SessionState$ExportBuilder.lambda$submit$2(SessionState.java:1299)
        at io.deephaven.server.session.SessionState$ExportObject.doExport(SessionState.java:847)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at io.deephaven.server.runner.DeephavenApiServerModule$ThreadFactory.lambda$newThread$0(DeephavenApiServerModule.java:162)
        at java.lang.Thread.run(Thread.java:829)
caused by:
java.lang.RuntimeException: Error compiling class io.deephaven.temp.c1359401156n730523876v55_0.GeneratedFilterKernel:
/io/deephaven/temp/c1359401156n730523876v55_0/GeneratedFilterKernel.java:87: error: cannot find symbol
            if (not(isNull(Batter))) {
                           ^
  symbol:   variable Batter
  location: class GeneratedFilterKernel
1 error

        at io.deephaven.compilertools.CompilerTools.maybeCreateClassHelper(CompilerTools.java:789)
        at io.deephaven.compilertools.CompilerTools.maybeCreateClass(CompilerTools.java:752)
        at io.deephaven.compilertools.CompilerTools.compileHelper(CompilerTools.java:510)
        at io.deephaven.compilertools.CompilerTools.compile(CompilerTools.java:469)
        at io.deephaven.compilertools.CompilerTools.compile(CompilerTools.java:424)
        at io.deephaven.engine.table.impl.select.ConditionFilter.generateFilterCode(ConditionFilter.java:392)
        at io.deephaven.engine.table.impl.select.AbstractConditionFilter.init(AbstractConditionFilter.java:197)
        at io.deephaven.engine.table.impl.QueryTable.lambda$whereInternal$28(QueryTable.java:1023)
        at io.deephaven.engine.table.impl.perf.QueryPerformanceRecorder.withNugget(QueryPerformanceRecorder.java:518)
        at io.deephaven.engine.table.impl.QueryTable.whereInternal(QueryTable.java:993)
        at io.deephaven.engine.table.impl.QueryTable.where(QueryTable.java:989)
        at io.deephaven.engine.table.impl.TableWithDefaults.where(TableWithDefaults.java:212)
        at io.deephaven.engine.table.impl.DeferredViewTable.getResultTableWithWhere(DeferredViewTable.java:103)
        at io.deephaven.engine.table.impl.DeferredViewTable.where(DeferredViewTable.java:73)
        at io.deephaven.engine.table.impl.TableWithDefaults.where(TableWithDefaults.java:218)
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:566)
        at org.codehaus.groovy.runtime.callsite.PlainObjectMetaMethodSite.doInvoke(PlainObjectMetaMethodSite.java:43)
        at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite$PojoCachedMethodSite.invoke(PojoMetaMethodSite.java:191)
        at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite.call(PojoMetaMethodSite.java:56)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:139)
        at io.deephaven.dynamic.Script_17.run(Script_17.groovy:43)
        at groovy.lang.GroovyShell.evaluate(GroovyShell.java:427)
        at groovy.lang.GroovyShell.evaluate(GroovyShell.java:461)
        at groovy.lang.GroovyShell.evaluate(GroovyShell.java:436)
        at io.deephaven.engine.util.GroovyDeephavenSession.evaluateCommand(GroovyDeephavenSession.java:228)
        at io.deephaven.engine.util.GroovyDeephavenSession.lambda$evaluate$2(GroovyDeephavenSession.java:251)
        at io.deephaven.util.locks.FunctionalLock.doLockedInterruptibly(FunctionalLock.java:49)
        at io.deephaven.engine.util.GroovyDeephavenSession.evaluate(GroovyDeephavenSession.java:251)
        at io.deephaven.engine.util.AbstractScriptSession.evaluateScript(AbstractScriptSession.java:166)
        at io.deephaven.engine.util.DelegatingScriptSession.evaluateScript(DelegatingScriptSession.java:81)
        at io.deephaven.engine.util.ScriptSession.evaluateScript(ScriptSession.java:106)
        at io.deephaven.server.console.ConsoleServiceGrpcImpl.lambda$executeCommand$8(ConsoleServiceGrpcImpl.java:188)
        at io.deephaven.server.session.SessionState$ExportBuilder.lambda$submit$2(SessionState.java:1299)
        at io.deephaven.server.session.SessionState$ExportObject.doExport(SessionState.java:847)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at io.deephaven.server.runner.DeephavenApiServerModule$ThreadFactory.lambda$newThread$0(DeephavenApiServerModule.java:162)
        at java.lang.Thread.run(Thread.java:829)
```

--- a/engine/table/src/main/java/io/deephaven/engine/table/impl/DeferredViewTable.java
+++ b/engine/table/src/main/java/io/deephaven/engine/table/impl/DeferredViewTable.java
@@ -10,6 +10,7 @@ import io.deephaven.base.verify.Assert;
 import io.deephaven.datastructures.util.CollectionUtil;
 import io.deephaven.engine.table.*;
 import io.deephaven.engine.table.impl.select.analyzers.SelectAndViewAnalyzer;
+import io.deephaven.engine.table.impl.select.MatchFilter;
 import io.deephaven.util.QueryConstants;
 import io.deephaven.engine.liveness.LivenessArtifact;
 import io.deephaven.engine.liveness.LivenessReferent;
@@ -18,6 +19,8 @@ import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
 import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 /**
@@ -181,30 +184,26 @@ public class DeferredViewTable extends RedefinableTable {
             }
         }
 
-        for (WhereFilter filter : filters) {
+        for (final WhereFilter filter : filters) {
             filter.init(definition);
-            final Set<String> usedColumns = new HashSet<>();
-            usedColumns.addAll(filter.getColumns());
-            usedColumns.addAll(filter.getColumnArrays());
 
-            if (usedColumns.stream().anyMatch(postViewColumns::contains)) {
+            final boolean isPostView = Stream.of(filter.getColumns(), filter.getColumnArrays())
+                    .flatMap(Collection::stream)
+                    .anyMatch(postViewColumns::contains);
+            if (isPostView) {
                 postViewFilters.add(filter);
                 continue;
             }
 
-            final Map<String, String> myRenames = new HashMap<>();
-            for (final String usedColumn : usedColumns) {
-                final String rename = renames.get(usedColumn);
-                if (rename != null) {
-                    myRenames.put(usedColumn, rename);
-                }
-            }
+            final Map<String, String> myRenames = Stream.of(filter.getColumns(), filter.getColumnArrays())
+                    .flatMap(Collection::stream)
+                    .filter(renames::containsKey)
+                    .collect(Collectors.toMap(Function.identity(), renames::get));
 
             if (myRenames.isEmpty()) {
                 preViewFilters.add(filter);
-            } else if (filter instanceof io.deephaven.engine.table.impl.select.MatchFilter) {
-                io.deephaven.engine.table.impl.select.MatchFilter matchFilter =
-                        (io.deephaven.engine.table.impl.select.MatchFilter) filter;
+            } else if (filter instanceof MatchFilter) {
+                MatchFilter matchFilter = (MatchFilter) filter;
                 Assert.assertion(myRenames.size() == 1, "Match Filters should only use one column!");
                 String newName = myRenames.get(matchFilter.getColumnName());
                 Assert.neqNull(newName, "newName");

--- a/engine/table/src/main/java/io/deephaven/engine/table/impl/DeferredViewTable.java
+++ b/engine/table/src/main/java/io/deephaven/engine/table/impl/DeferredViewTable.java
@@ -18,6 +18,7 @@ import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
 import java.util.*;
+import java.util.function.Supplier;
 import java.util.stream.Stream;
 
 /**
@@ -148,28 +149,55 @@ public class DeferredViewTable extends RedefinableTable {
 
         for (WhereFilter filter : filters) {
             filter.init(definition);
-            ArrayList<String> usedColumns = new ArrayList<>();
+            final Set<String> usedColumns = new HashSet<>();
             usedColumns.addAll(filter.getColumns());
             usedColumns.addAll(filter.getColumnArrays());
 
-            Map<String, String> renames = new HashMap<>();
             boolean postFilter = false;
+            Map<String, String> renames = new HashMap<>();
+            final Map<String, Set<String>> reverseLookup = new HashMap<>();
 
-            for (String column : usedColumns) {
-                for (SelectColumn selectColumn : deferredViewColumns) {
-                    if (selectColumn.getName().equals(column)) {
-                        if (selectColumn instanceof SwitchColumn) {
-                            selectColumn.initDef(tableReference.getDefinition().getColumnNameMap());
-                            selectColumn = ((SwitchColumn) selectColumn).getRealColumn();
-                        }
+            // determine if filter has any dependencies on deferred columns
+            for (int ii = deferredViewColumns.length - 1; ii >= 0; --ii) {
+                // traverse columns in reverse order to handle scenarios where a deferred column is renamed
+                SelectColumn selectColumn = deferredViewColumns[ii];
+                final String columnName = selectColumn.getName();
+                if (selectColumn.isRetain()) {
+                    continue;
+                }
+                if (!usedColumns.contains(columnName) && !reverseLookup.containsKey(columnName)) {
+                    continue;
+                }
+                if (selectColumn instanceof SwitchColumn) {
+                    selectColumn.initDef(tableReference.getDefinition().getColumnNameMap());
+                    selectColumn = ((SwitchColumn) selectColumn).getRealColumn();
+                }
 
-                        // we need to rename this column
-                        if (selectColumn instanceof SourceColumn) {
-                            // this is a rename of the getSourceName to the innerName
-                            renames.put(selectColumn.getName(), ((SourceColumn) selectColumn).getSourceName());
-                        } else {
-                            postFilter = true;
-                            break;
+                final String innerName;
+                if (selectColumn instanceof SourceColumn) {
+                    innerName = ((SourceColumn) selectColumn).getSourceName();
+                } else {
+                    // this is a deferred column
+                    postFilter = true;
+                    break;
+                }
+
+                // we need to rename this column
+                renames.put(columnName, innerName);
+                reverseLookup.putIfAbsent(innerName, new HashSet<>());
+
+                // check if this is an explicit rename
+                if (usedColumns.remove(columnName)) {
+                    reverseLookup.get(innerName).add(columnName);
+                }
+
+                // apply this rename to any other deferred columns that depend on this one
+                if (!selectColumn.isRetain()) {
+                    final Set<String> multiLevelRenames = reverseLookup.remove(columnName);
+                    if (multiLevelRenames != null) {
+                        for (final String resultColumnName : multiLevelRenames) {
+                            renames.put(resultColumnName, innerName);
+                            reverseLookup.get(innerName).add(resultColumnName);
                         }
                     }
                 }
@@ -177,24 +205,20 @@ public class DeferredViewTable extends RedefinableTable {
 
             if (postFilter) {
                 postViewFilters.add(filter);
+            } else if (renames.isEmpty()) {
+                preViewFilters.add(filter);
+            } else if (filter instanceof io.deephaven.engine.table.impl.select.MatchFilter) {
+                io.deephaven.engine.table.impl.select.MatchFilter matchFilter =
+                        (io.deephaven.engine.table.impl.select.MatchFilter) filter;
+                Assert.assertion(renames.size() == 1, "Match Filters should only use one column!");
+                String newName = renames.get(matchFilter.getColumnName());
+                Assert.neqNull(newName, "newName");
+                preViewFilters.add(matchFilter.renameFilter(newName));
+            } else if (filter instanceof ConditionFilter) {
+                ConditionFilter conditionFilter = (ConditionFilter) filter;
+                preViewFilters.add(conditionFilter.renameFilter(renames));
             } else {
-                if (!renames.isEmpty()) {
-                    if (filter instanceof io.deephaven.engine.table.impl.select.MatchFilter) {
-                        io.deephaven.engine.table.impl.select.MatchFilter matchFilter =
-                                (io.deephaven.engine.table.impl.select.MatchFilter) filter;
-                        Assert.assertion(renames.size() == 1, "Match Filters should only use one column!");
-                        String newName = renames.get(matchFilter.getColumnName());
-                        Assert.neqNull(newName, "newName");
-                        preViewFilters.add(matchFilter.renameFilter(newName));
-                    } else if (filter instanceof ConditionFilter) {
-                        ConditionFilter conditionFilter = (ConditionFilter) filter;
-                        preViewFilters.add(conditionFilter.renameFilter(renames));
-                    } else {
-                        postViewFilters.add(filter);
-                    }
-                } else {
-                    preViewFilters.add(filter);
-                }
+                postViewFilters.add(filter);
             }
         }
 

--- a/engine/table/src/main/java/io/deephaven/engine/table/impl/DeferredViewTable.java
+++ b/engine/table/src/main/java/io/deephaven/engine/table/impl/DeferredViewTable.java
@@ -18,7 +18,6 @@ import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
 import java.util.*;
-import java.util.function.Supplier;
 import java.util.stream.Stream;
 
 /**
@@ -147,76 +146,84 @@ public class DeferredViewTable extends RedefinableTable {
         ArrayList<WhereFilter> preViewFilters = new ArrayList<>();
         ArrayList<WhereFilter> postViewFilters = new ArrayList<>();
 
+        final Set<String> postViewColumns = new HashSet<>();
+        final Map<String, String> renames = new HashMap<>();
+
+        // keep track of renames already applied, so that we can detect the case when a rename is renamed
+        final Map<String, Set<String>> reverseLookup = new HashMap<>();
+
+        // determine if filter has any dependencies on deferred columns
+        for (int ii = deferredViewColumns.length - 1; ii >= 0; --ii) {
+            // traverse columns in reverse order to handle scenarios where a deferred column is renamed
+            SelectColumn selectColumn = deferredViewColumns[ii];
+            final String columnName = selectColumn.getName();
+            if (selectColumn.isRetain()) {
+                continue;
+            }
+            if (selectColumn instanceof SwitchColumn) {
+                selectColumn = ((SwitchColumn) selectColumn).getRealColumn();
+            }
+
+            if (selectColumn instanceof SourceColumn) {
+                // this is a renamed column
+                final String innerName = ((SourceColumn) selectColumn).getSourceName();
+                final Set<String> myOuterNames = reverseLookup.computeIfAbsent(innerName, name -> new HashSet<>());
+
+                myOuterNames.add(columnName);
+                renames.put(columnName, innerName);
+
+                final Set<String> multiLevelRenames = reverseLookup.remove(columnName);
+                if (multiLevelRenames != null) {
+                    myOuterNames.addAll(multiLevelRenames);
+                    for (final String resultColumnName : multiLevelRenames) {
+                        renames.put(resultColumnName, innerName);
+                    }
+                }
+            } else {
+                // this is a deferred column
+                postViewColumns.add(columnName);
+
+                final Set<String> multiLevelRenames = reverseLookup.remove(columnName);
+                if (multiLevelRenames != null) {
+                    for (final String outerName : multiLevelRenames) {
+                        renames.remove(outerName);
+                        postViewColumns.add(outerName);
+                    }
+                }
+            }
+        }
+
         for (WhereFilter filter : filters) {
             filter.init(definition);
             final Set<String> usedColumns = new HashSet<>();
             usedColumns.addAll(filter.getColumns());
             usedColumns.addAll(filter.getColumnArrays());
 
-            boolean postFilter = false;
-            Map<String, String> renames = new HashMap<>();
-            final Map<String, Set<String>> reverseLookup = new HashMap<>();
+            if (usedColumns.stream().anyMatch(postViewColumns::contains)) {
+                postViewFilters.add(filter);
+                continue;
+            }
 
-            // determine if filter has any dependencies on deferred columns
-            for (int ii = deferredViewColumns.length - 1; ii >= 0; --ii) {
-                // traverse columns in reverse order to handle scenarios where a deferred column is renamed
-                SelectColumn selectColumn = deferredViewColumns[ii];
-                final String columnName = selectColumn.getName();
-                if (selectColumn.isRetain()) {
-                    continue;
-                }
-                if (!usedColumns.contains(columnName) && !reverseLookup.containsKey(columnName)) {
-                    continue;
-                }
-                if (selectColumn instanceof SwitchColumn) {
-                    selectColumn.initDef(tableReference.getDefinition().getColumnNameMap());
-                    selectColumn = ((SwitchColumn) selectColumn).getRealColumn();
-                }
-
-                final String innerName;
-                if (selectColumn instanceof SourceColumn) {
-                    innerName = ((SourceColumn) selectColumn).getSourceName();
-                } else {
-                    // this is a deferred column
-                    postFilter = true;
-                    break;
-                }
-
-                // we need to rename this column
-                renames.put(columnName, innerName);
-                reverseLookup.putIfAbsent(innerName, new HashSet<>());
-
-                // check if this is an explicit rename
-                if (usedColumns.remove(columnName)) {
-                    reverseLookup.get(innerName).add(columnName);
-                }
-
-                // apply this rename to any other deferred columns that depend on this one
-                if (!selectColumn.isRetain()) {
-                    final Set<String> multiLevelRenames = reverseLookup.remove(columnName);
-                    if (multiLevelRenames != null) {
-                        for (final String resultColumnName : multiLevelRenames) {
-                            renames.put(resultColumnName, innerName);
-                            reverseLookup.get(innerName).add(resultColumnName);
-                        }
-                    }
+            final Map<String, String> myRenames = new HashMap<>();
+            for (final String usedColumn : usedColumns) {
+                final String rename = renames.get(usedColumn);
+                if (rename != null) {
+                    myRenames.put(usedColumn, rename);
                 }
             }
 
-            if (postFilter) {
-                postViewFilters.add(filter);
-            } else if (renames.isEmpty()) {
+            if (myRenames.isEmpty()) {
                 preViewFilters.add(filter);
             } else if (filter instanceof io.deephaven.engine.table.impl.select.MatchFilter) {
                 io.deephaven.engine.table.impl.select.MatchFilter matchFilter =
                         (io.deephaven.engine.table.impl.select.MatchFilter) filter;
-                Assert.assertion(renames.size() == 1, "Match Filters should only use one column!");
-                String newName = renames.get(matchFilter.getColumnName());
+                Assert.assertion(myRenames.size() == 1, "Match Filters should only use one column!");
+                String newName = myRenames.get(matchFilter.getColumnName());
                 Assert.neqNull(newName, "newName");
                 preViewFilters.add(matchFilter.renameFilter(newName));
             } else if (filter instanceof ConditionFilter) {
                 ConditionFilter conditionFilter = (ConditionFilter) filter;
-                preViewFilters.add(conditionFilter.renameFilter(renames));
+                preViewFilters.add(conditionFilter.renameFilter(myRenames));
             } else {
                 postViewFilters.add(filter);
             }

--- a/engine/table/src/main/java/io/deephaven/engine/table/impl/DeferredViewTable.java
+++ b/engine/table/src/main/java/io/deephaven/engine/table/impl/DeferredViewTable.java
@@ -149,14 +149,9 @@ public class DeferredViewTable extends RedefinableTable {
         final Set<String> postViewColumns = new HashSet<>();
         final Map<String, String> renames = new HashMap<>();
 
-        // keep track of renames already applied, so that we can detect the case when a rename is renamed
-        final Map<String, Set<String>> reverseLookup = new HashMap<>();
-
-        // determine if filter has any dependencies on deferred columns
-        for (int ii = deferredViewColumns.length - 1; ii >= 0; --ii) {
-            // traverse columns in reverse order to handle scenarios where a deferred column is renamed
-            SelectColumn selectColumn = deferredViewColumns[ii];
-            final String columnName = selectColumn.getName();
+        // prepare data structures needed to determine if filter can be applied pre-view or if it must be post-view
+        for (SelectColumn selectColumn : deferredViewColumns) {
+            final String outerName = selectColumn.getName();
             if (selectColumn.isRetain()) {
                 continue;
             }
@@ -164,32 +159,25 @@ public class DeferredViewTable extends RedefinableTable {
                 selectColumn = ((SwitchColumn) selectColumn).getRealColumn();
             }
 
+            // This column is being defined; whatever we currently believe might be wrong
+            renames.remove(outerName);
+            postViewColumns.remove(outerName);
+
             if (selectColumn instanceof SourceColumn) {
                 // this is a renamed column
                 final String innerName = ((SourceColumn) selectColumn).getSourceName();
-                final Set<String> myOuterNames = reverseLookup.computeIfAbsent(innerName, name -> new HashSet<>());
+                final String sourceName = renames.getOrDefault(innerName, innerName);
 
-                myOuterNames.add(columnName);
-                renames.put(columnName, innerName);
-
-                final Set<String> multiLevelRenames = reverseLookup.remove(columnName);
-                if (multiLevelRenames != null) {
-                    myOuterNames.addAll(multiLevelRenames);
-                    for (final String resultColumnName : multiLevelRenames) {
-                        renames.put(resultColumnName, innerName);
-                    }
+                if (postViewColumns.contains(sourceName)) {
+                    // but it is renamed to a deferred column
+                    postViewColumns.add(outerName);
+                } else {
+                    // this is renamed to a real source column
+                    renames.put(outerName, sourceName);
                 }
             } else {
                 // this is a deferred column
-                postViewColumns.add(columnName);
-
-                final Set<String> multiLevelRenames = reverseLookup.remove(columnName);
-                if (multiLevelRenames != null) {
-                    for (final String outerName : multiLevelRenames) {
-                        renames.remove(outerName);
-                        postViewColumns.add(outerName);
-                    }
-                }
+                postViewColumns.add(outerName);
             }
         }
 

--- a/engine/table/src/main/java/io/deephaven/engine/table/impl/select/AbstractConditionFilter.java
+++ b/engine/table/src/main/java/io/deephaven/engine/table/impl/select/AbstractConditionFilter.java
@@ -33,6 +33,7 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.function.BiConsumer;
+import java.util.stream.Collectors;
 
 import static io.deephaven.engine.table.impl.select.DhFormulaColumn.COLUMN_SUFFIX;
 
@@ -65,20 +66,18 @@ public abstract class AbstractConditionFilter extends WhereFilterImpl {
 
     @Override
     public List<String> getColumns() {
-        final ArrayList<String> result = new ArrayList<>(usedColumns.size());
-        for (final String usedColumn : usedColumns) {
-            result.add(outerToInnerNames.getOrDefault(usedColumn, usedColumn));
-        }
-        return result;
+        return usedColumns.stream()
+                .map(name -> outerToInnerNames.getOrDefault(name, name))
+                .distinct()
+                .collect(Collectors.toList());
     }
 
     @Override
     public List<String> getColumnArrays() {
-        final ArrayList<String> result = new ArrayList<>(usedColumnArrays.size());
-        for (final String usedColumn : usedColumnArrays) {
-            result.add(outerToInnerNames.getOrDefault(usedColumn, usedColumn));
-        }
-        return result;
+        return usedColumnArrays.stream()
+                .map(name -> outerToInnerNames.getOrDefault(name, name))
+                .distinct()
+                .collect(Collectors.toList());
     }
 
     @Override
@@ -116,7 +115,7 @@ public abstract class AbstractConditionFilter extends WhereFilterImpl {
                 possibleVariables.put(columnName, column.getDataType());
                 possibleVariables.put(columnName + COLUMN_SUFFIX, vectorType);
 
-                Class<?> compType = column.getComponentType();
+                final Class<?> compType = column.getComponentType();
                 if (compType != null && !compType.isPrimitive()) {
                     possibleVariableParameterizedTypes.put(columnName, new Class[] {compType});
                 }

--- a/engine/table/src/main/java/io/deephaven/engine/table/impl/select/AbstractConditionFilter.java
+++ b/engine/table/src/main/java/io/deephaven/engine/table/impl/select/AbstractConditionFilter.java
@@ -32,13 +32,13 @@ import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.function.BiConsumer;
 
 import static io.deephaven.engine.table.impl.select.DhFormulaColumn.COLUMN_SUFFIX;
 
 public abstract class AbstractConditionFilter extends WhereFilterImpl {
     private static final Logger log = LoggerFactory.getLogger(AbstractConditionFilter.class);
     final Map<String, String> outerToInnerNames;
-    final Map<String, String> innerToOuterNames;
     @NotNull
     protected final String formula;
     List<String> usedColumns;
@@ -55,27 +55,30 @@ public abstract class AbstractConditionFilter extends WhereFilterImpl {
         this.formula = formula;
         this.unboxArguments = unboxArguments;
         this.outerToInnerNames = Collections.emptyMap();
-        this.innerToOuterNames = Collections.emptyMap();
     }
 
     protected AbstractConditionFilter(@NotNull String formula, Map<String, String> renames, boolean unboxArguments) {
         this.formula = formula;
         this.outerToInnerNames = renames;
         this.unboxArguments = unboxArguments;
-        this.innerToOuterNames = new HashMap<>();
-        for (Map.Entry<String, String> outerInnerEntry : outerToInnerNames.entrySet()) {
-            innerToOuterNames.put(outerInnerEntry.getValue(), outerInnerEntry.getKey());
-        }
     }
 
     @Override
     public List<String> getColumns() {
-        return usedColumns;
+        final ArrayList<String> result = new ArrayList<>(usedColumns.size());
+        for (final String usedColumn : usedColumns) {
+            result.add(outerToInnerNames.getOrDefault(usedColumn, usedColumn));
+        }
+        return result;
     }
 
     @Override
     public List<String> getColumnArrays() {
-        return usedColumnArrays;
+        final ArrayList<String> result = new ArrayList<>(usedColumnArrays.size());
+        for (final String usedColumn : usedColumnArrays) {
+            result.add(outerToInnerNames.getOrDefault(usedColumn, usedColumn));
+        }
+        return result;
     }
 
     @Override
@@ -107,15 +110,13 @@ public abstract class AbstractConditionFilter extends WhereFilterImpl {
                 }
             }
 
-            Class<?> compType;
-            for (ColumnDefinition<?> column : tableDefinition.getColumns()) {
+            final BiConsumer<String, ColumnDefinition<?>> createColumnMappings = (columnName, column) -> {
                 final Class<?> vectorType = DhFormulaColumn.getVectorType(column.getDataType());
-                final String columnName = innerToOuterNames.getOrDefault(column.getName(), column.getName());
 
                 possibleVariables.put(columnName, column.getDataType());
                 possibleVariables.put(columnName + COLUMN_SUFFIX, vectorType);
 
-                compType = column.getComponentType();
+                Class<?> compType = column.getComponentType();
                 if (compType != null && !compType.isPrimitive()) {
                     possibleVariableParameterizedTypes.put(columnName, new Class[] {compType});
                 }
@@ -123,6 +124,17 @@ public abstract class AbstractConditionFilter extends WhereFilterImpl {
                     possibleVariableParameterizedTypes.put(columnName + COLUMN_SUFFIX,
                             new Class[] {column.getDataType()});
                 }
+            };
+
+            // By default all columns are available to the formula
+            for (final ColumnDefinition<?> column : tableDefinition.getColumns()) {
+                createColumnMappings.accept(column.getName(), column);
+            }
+            // Overwrite any existing column mapping using the provided renames.
+            for (final Map.Entry<String, String> entry : outerToInnerNames.entrySet()) {
+                final String columnName = entry.getKey();
+                final ColumnDefinition<?> column = tableDefinition.getColumn(entry.getValue());
+                createColumnMappings.accept(columnName, column);
             }
 
             log.debug("Expression (before) : " + formula);
@@ -149,11 +161,13 @@ public abstract class AbstractConditionFilter extends WhereFilterImpl {
             for (String variable : result.getVariablesUsed()) {
                 final String columnToFind = outerToInnerNames.getOrDefault(variable, variable);
                 final String arrayColumnToFind;
+                final String arrayColumnOuterName;
                 if (variable.endsWith(COLUMN_SUFFIX)) {
-                    final String originalName = variable.substring(0, variable.length() - COLUMN_SUFFIX.length());
-                    arrayColumnToFind = outerToInnerNames.getOrDefault(originalName, originalName);
+                    arrayColumnOuterName = variable.substring(0, variable.length() - COLUMN_SUFFIX.length());
+                    arrayColumnToFind = outerToInnerNames.getOrDefault(arrayColumnOuterName, arrayColumnOuterName);
                 } else {
                     arrayColumnToFind = null;
+                    arrayColumnOuterName = null;
                 }
 
                 if (variable.equals("i")) {
@@ -163,9 +177,9 @@ public abstract class AbstractConditionFilter extends WhereFilterImpl {
                 } else if (variable.equals("k")) {
                     usesK = true;
                 } else if (tableDefinition.getColumn(columnToFind) != null) {
-                    usedColumns.add(columnToFind);
+                    usedColumns.add(variable);
                 } else if (arrayColumnToFind != null && tableDefinition.getColumn(arrayColumnToFind) != null) {
-                    usedColumnArrays.add(arrayColumnToFind);
+                    usedColumnArrays.add(arrayColumnOuterName);
                 } else if (possibleParams.containsKey(variable)) {
                     paramsList.add(possibleParams.get(variable));
                 }

--- a/engine/table/src/main/java/io/deephaven/engine/table/impl/select/ConditionFilter.java
+++ b/engine/table/src/main/java/io/deephaven/engine/table/impl/select/ConditionFilter.java
@@ -46,7 +46,7 @@ public class ConditionFilter extends AbstractConditionFilter {
 
     public static final int CHUNK_SIZE = 4096;
     private Class<?> filterKernelClass = null;
-    private List<Pair<String, Class>> usedInputs; // that is columns and special variables
+    private List<Pair<String, Class<?>>> usedInputs; // that is columns and special variables
     private String classBody;
     private Filter filter = null;
     private boolean filterValidForCopy = true;
@@ -368,8 +368,6 @@ public class ConditionFilter extends AbstractConditionFilter {
                 return resultBuilder.build();
             }
         }
-
-
     }
 
     private static String toTitleCase(String input) {
@@ -378,7 +376,7 @@ public class ConditionFilter extends AbstractConditionFilter {
 
     @Override
     protected void generateFilterCode(TableDefinition tableDefinition, DateTimeUtils.Result timeConversionResult,
-            QueryLanguageParser.Result result) throws MalformedURLException, ClassNotFoundException {
+            QueryLanguageParser.Result result) {
         final StringBuilder classBody = getClassBody(tableDefinition, timeConversionResult, result);
         if (classBody == null)
             return;
@@ -390,12 +388,14 @@ public class ConditionFilter extends AbstractConditionFilter {
                     paramClasses.add(cls);
                 }
             };
-            for (final String usedColumn : usedColumns) {
+            for (String usedColumn : usedColumns) {
+                usedColumn = outerToInnerNames.getOrDefault(usedColumn, usedColumn);
                 final ColumnDefinition<?> column = tableDefinition.getColumn(usedColumn);
                 addParamClass.accept(column.getDataType());
                 addParamClass.accept(column.getComponentType());
             }
-            for (final String usedColumn : usedColumnArrays) {
+            for (String usedColumn : usedColumnArrays) {
+                usedColumn = outerToInnerNames.getOrDefault(usedColumn, usedColumn);
                 final ColumnDefinition<?> column = tableDefinition.getColumn(usedColumn);
                 addParamClass.accept(column.getDataType());
                 addParamClass.accept(column.getComponentType());
@@ -420,8 +420,9 @@ public class ConditionFilter extends AbstractConditionFilter {
         }
         usedInputs = new ArrayList<>();
         for (String usedColumn : usedColumns) {
-            final ColumnDefinition column = tableDefinition.getColumn(usedColumn);
-            final Class columnType = column.getDataType();
+            final String innerName = outerToInnerNames.getOrDefault(usedColumn, usedColumn);
+            final ColumnDefinition<?> column = tableDefinition.getColumn(innerName);
+            final Class<?> columnType = column.getDataType();
             usedInputs.add(new Pair<>(usedColumn, columnType));
         }
         if (usesI) {
@@ -442,10 +443,15 @@ public class ConditionFilter extends AbstractConditionFilter {
                 .append(FilterKernel.class.getCanonicalName()).append("<FilterKernel.Context>{\n");
         classBody.append("\n").append(timeConversionResult.getInstanceVariablesString()).append("\n");
         final Indenter indenter = new Indenter();
-        for (QueryScopeParam param : params) {
+        for (QueryScopeParam<?> param : params) {
             /*
-             * adding context param fields like: "            final int p1;\n" + "            final float p2;\n" +
-             * "            final String p3;\n" +
+             * @formatter:off
+             * adding context param fields like:
+             *
+             * final int p1;
+             * final float p2;
+             * final String p3;
+             * @formatter:on
              */
             classBody.append(indenter).append("private final ")
                     .append(QueryScopeParamTypeUtil.getPrimitiveTypeNameIfAvailable(param.getValue()))
@@ -454,12 +460,13 @@ public class ConditionFilter extends AbstractConditionFilter {
         if (!usedColumnArrays.isEmpty()) {
             classBody.append(indenter).append("// Array Column Variables\n");
             for (String columnName : usedColumnArrays) {
-                final ColumnDefinition column = tableDefinition.getColumn(columnName);
+                final String innerColumnName = outerToInnerNames.getOrDefault(columnName, columnName);
+                final ColumnDefinition<?> column = tableDefinition.getColumn(innerColumnName);
                 if (column == null) {
-                    throw new RuntimeException("Column \"" + columnName + "\" doesn't exist in this table");
+                    throw new RuntimeException("Column \"" + innerColumnName + "\" doesn't exist in this table");
                 }
-                final Class dataType = column.getDataType();
-                final Class columnType = DhFormulaColumn.getVectorType(dataType);
+                final Class<?> dataType = column.getDataType();
+                final Class<?> columnType = DhFormulaColumn.getVectorType(dataType);
 
                 /*
                  * Adding array column fields.
@@ -467,7 +474,7 @@ public class ConditionFilter extends AbstractConditionFilter {
                 classBody.append(indenter).append("private final ").append(columnType.getCanonicalName())
                         .append(TypeUtils.isConvertibleToPrimitive(dataType) ? ""
                                 : "<" + dataType.getCanonicalName() + ">")
-                        .append(" ").append(column.getName()).append(COLUMN_SUFFIX).append(";\n");
+                        .append(" ").append(columnName).append(COLUMN_SUFFIX).append(";\n");
             }
             classBody.append("\n");
         }
@@ -476,7 +483,7 @@ public class ConditionFilter extends AbstractConditionFilter {
                 .append("public $CLASSNAME$(Table __table, RowSet __fullSet, QueryScopeParam... __params) {\n");
         indenter.increaseLevel();
         for (int i = 0; i < params.length; i++) {
-            final QueryScopeParam param = params[i];
+            final QueryScopeParam<?> param = params[i];
             /*
              * @formatter:off
              * Initializing context parameters:
@@ -496,12 +503,13 @@ public class ConditionFilter extends AbstractConditionFilter {
             classBody.append("\n");
             classBody.append(indenter).append("// Array Column Variables\n");
             for (String columnName : usedColumnArrays) {
-                final ColumnDefinition column = tableDefinition.getColumn(columnName);
+                final String innerColumnName = outerToInnerNames.getOrDefault(columnName, columnName);
+                final ColumnDefinition<?> column = tableDefinition.getColumn(innerColumnName);
                 if (column == null) {
-                    throw new RuntimeException("Column \"" + columnName + "\" doesn't exist in this table");
+                    throw new RuntimeException("Column \"" + innerColumnName + "\" doesn't exist in this table");
                 }
-                final Class dataType = column.getDataType();
-                final Class columnType = DhFormulaColumn.getVectorType(dataType);
+                final Class<?> dataType = column.getDataType();
+                final Class<?> columnType = DhFormulaColumn.getVectorType(dataType);
 
                 final String arrayType = columnType.getCanonicalName().replace(
                         "io.deephaven.vector",
@@ -510,8 +518,8 @@ public class ConditionFilter extends AbstractConditionFilter {
                 /*
                  * Adding array column fields.
                  */
-                classBody.append(indenter).append(column.getName()).append(COLUMN_SUFFIX).append(" = new ")
-                        .append(arrayType).append("(__table.getColumnSource(\"").append(columnName)
+                classBody.append(indenter).append(columnName).append(COLUMN_SUFFIX).append(" = new ")
+                        .append(arrayType).append("(__table.getColumnSource(\"").append(innerColumnName)
                         .append("\"), __fullSet);\n");
             }
         }
@@ -528,7 +536,7 @@ public class ConditionFilter extends AbstractConditionFilter {
                 "public LongChunk<OrderedRowKeys> filter(Context __context, LongChunk<OrderedRowKeys> __indices, Chunk... __inputChunks) {\n");
         indenter.increaseLevel();
         for (int i = 0; i < usedInputs.size(); i++) {
-            final Class columnType = usedInputs.get(i).second;
+            final Class<?> columnType = usedInputs.get(i).second;
             final String chunkType;
             if (columnType.isPrimitive() && columnType != boolean.class) {
                 chunkType = toTitleCase(columnType.getSimpleName()) + "Chunk";
@@ -544,8 +552,8 @@ public class ConditionFilter extends AbstractConditionFilter {
                 "for (int __my_i__ = 0; __my_i__ < __size; __my_i__++) {\n");
         indenter.increaseLevel();
         for (int i = 0; i < usedInputs.size(); i++) {
-            final Pair<String, Class> usedInput = usedInputs.get(i);
-            final Class columnType = usedInput.second;
+            final Pair<String, Class<?>> usedInput = usedInputs.get(i);
+            final Class<?> columnType = usedInput.second;
             final String canonicalName = columnType.getCanonicalName();
             classBody.append(indenter).append("final ").append(canonicalName).append(" ").append(usedInput.first)
                     .append(" =  (").append(canonicalName).append(")__columnChunk").append(i)
@@ -569,7 +577,9 @@ public class ConditionFilter extends AbstractConditionFilter {
             final FilterKernel<?> filterKernel = (FilterKernel<?>) filterKernelClass
                     .getConstructor(Table.class, RowSet.class, QueryScopeParam[].class)
                     .newInstance(table, fullSet, (Object) params);
-            final String[] columnNames = usedInputs.stream().map(p -> p.first).toArray(String[]::new);
+            final String[] columnNames = usedInputs.stream()
+                    .map(p -> outerToInnerNames.getOrDefault(p.first, p.first))
+                    .toArray(String[]::new);
             filter = new ChunkFilter(filterKernel, columnNames, CHUNK_SIZE);
             // note this filter is not valid for use in other contexts, as it captures references from the source table
             filterValidForCopy = false;

--- a/engine/table/src/main/java/io/deephaven/engine/table/impl/select/ConditionFilter.java
+++ b/engine/table/src/main/java/io/deephaven/engine/table/impl/select/ConditionFilter.java
@@ -31,7 +31,6 @@ import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
 import java.lang.reflect.InvocationTargetException;
-import java.net.MalformedURLException;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;

--- a/engine/table/src/test/java/io/deephaven/engine/table/impl/PartitionedTableTest.java
+++ b/engine/table/src/test/java/io/deephaven/engine/table/impl/PartitionedTableTest.java
@@ -822,7 +822,7 @@ public class PartitionedTableTest extends RefreshingTableTestCase {
                         .captureQueryLibrary()
                         .newQueryScope()
                         .build().open()) {
-                    ;
+
                     ExecutionContext.getContext().getQueryScope().putParam("queryScopeVar", "queryScopeValue");
                     ExecutionContext.getContext().getQueryScope().putParam("queryScopeFilter", 50000);
 

--- a/extensions/parquet/table/src/test/java/io/deephaven/parquet/table/TestParquetTools.java
+++ b/extensions/parquet/table/src/test/java/io/deephaven/parquet/table/TestParquetTools.java
@@ -428,7 +428,8 @@ public class TestParquetTools {
                 t -> t.updateView("Y = X").where("(X_[ii] + Y_[ii]) % 2 == 0"));
     }
 
-    @Test public void testMultipleRenamesWithSameOuterName() {
+    @Test
+    public void testMultipleRenamesWithSameOuterName() {
         testWriteRead(emptyTable(10).update("X = ii", "Y = ii", "Z = ii % 3"),
                 t -> t.updateView("Y = Z", "Y = X").where("Y % 2 == 0"));
     }

--- a/extensions/parquet/table/src/test/java/io/deephaven/parquet/table/TestParquetTools.java
+++ b/extensions/parquet/table/src/test/java/io/deephaven/parquet/table/TestParquetTools.java
@@ -428,6 +428,11 @@ public class TestParquetTools {
                 t -> t.updateView("Y = X").where("(X_[ii] + Y_[ii]) % 2 == 0"));
     }
 
+    @Test public void testMultipleRenamesWithSameOuterName() {
+        testWriteRead(emptyTable(10).update("X = ii", "Y = ii", "Z = ii % 3"),
+                t -> t.updateView("Y = Z", "Y = X").where("Y % 2 == 0"));
+    }
+
     private void testWriteRead(Table source, Function<Table, Table> transform) {
         final File f2w = new File(testRoot, "testWriteRead.parquet");
         ParquetTools.writeTable(source, f2w);

--- a/extensions/parquet/table/src/test/java/io/deephaven/parquet/table/TestParquetTools.java
+++ b/extensions/parquet/table/src/test/java/io/deephaven/parquet/table/TestParquetTools.java
@@ -404,6 +404,30 @@ public class TestParquetTools {
                 t -> t.view("T = X_[i-1]"));
     }
 
+    @Test
+    public void testMultipleIdenticalRenames() {
+        testWriteRead(emptyTable(10).update("X = `` + ii"),
+                t -> t.updateView("A = X", "B = X").where("(A + B).length() > 0"));
+        testWriteRead(emptyTable(10).update("X = `` + ii"),
+                t -> t.updateView("A = X", "B = X").where("(A_[ii] + B_[ii]).length() > 0"));
+    }
+
+    @Test
+    public void testChangedThenRenamed() {
+        testWriteRead(emptyTable(10).update("X = ii", "Y = ii"),
+                t -> t.updateView("Y = ii * 2", "X = Y").where("X % 2 == 0"));
+        testWriteRead(emptyTable(10).update("X = ii", "Y = ii"),
+                t -> t.updateView("Y = ii * 2", "X = Y").where("X_[ii] % 2 == 0"));
+    }
+
+    @Test
+    public void testOverloadAsRename() {
+        testWriteRead(emptyTable(10).update("X = ii"),
+                t -> t.updateView("Y = X").where("(X + Y) % 2 == 0"));
+        testWriteRead(emptyTable(10).update("X = ii"),
+                t -> t.updateView("Y = X").where("(X_[ii] + Y_[ii]) % 2 == 0"));
+    }
+
     private void testWriteRead(Table source, Function<Table, Table> transform) {
         final File f2w = new File(testRoot, "testWriteRead.parquet");
         ParquetTools.writeTable(source, f2w);

